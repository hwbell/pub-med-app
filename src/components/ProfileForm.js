import React from 'react';
import '../App.scss';
import 'bootstrap/dist/css/bootstrap.css';

// components
import { Button, Modal, ModalHeader, ModalBody, ModalFooter, Input, Form } from 'reactstrap';

// ******************************************************************************
class ProfileForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      profileInfo: {}
    }

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(e) {

    let { name, value } = e.target;

    // use the target name as the key for the value in profileInfo
    let obj = this.state.profileInfo;
    obj[`${name}`] = value;

    this.setState({ profileInfo: obj }, () => {
      // console.log(this.state.user)
    })
  }

  handleSubmit() {
    this.props.handleSubmitUser(this.state.profileInfo);

    // wipe the state - we only need it before it is sent to the server
    // this.setState({
    // profileInfo: {}
    // })
  }

  render() {

    return (
      <Modal centered={true}
        isOpen={this.props.isVisible}
        toggle={this.props.toggle}>

        <ModalHeader>Add some info to your profile</ModalHeader>
        <ModalBody>

          <Form style={styles.form} onSubmit={this.handleSubmit}>

            <p style={{ color: 'black' }}>About</p>
            <Input style={styles.inputLarge} type="textarea" name="about"
              maxLength={200}
              placeholder="Say something about yourself!"
              onChange={(e) => this.handleChange(e)}
            />

            <p style={{ color: 'black' }}>Research</p>
            <Input style={styles.input} type="text" name="research"
              maxLength={50}
              placeholder="What is your area of expertise?"
              onChange={(e) => this.handleChange(e)}
            />

            <p style={{ color: 'black' }}>Affiliations</p>
            <Input style={styles.input} type="text" name="affiliations"
              maxLength={50}
              placeholder="Where do you conduct your research?"
              onChange={(e) => this.handleChange(e)}
            />

            <p style={{ color: 'black' }}>Interests</p>
            <Input style={styles.input} type="text" name="interests"
              maxLength={50}
              placeholder="What are your interests outside of research?"
              onChange={(e) => this.handleChange(e)}
            />

          </Form>

        </ModalBody>
        <ModalFooter>
          <Button color="primary" size="sm" onClick={this.handleSubmit}>save</Button>
          <Button color="secondary" size="sm" onClick={this.props.toggle}>cancel</Button>
        </ModalFooter>

      </Modal>
    );
  }
}

const styles = {
  input: {
    marginBottom: '20px',
    textAlign: 'left'
  },
  inputLarge: {
    marginBottom: '20px',
    height: '140px',
    textAlign: 'left',
    paddingTop: 0
  }
}

export default ProfileForm;
