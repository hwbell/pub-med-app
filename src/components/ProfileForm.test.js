import React from 'react';
import ProfileForm from './ProfileForm';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

// stub for toggling
const toggleStub = jest.fn();
const confirmStub = jest.fn();

const someProps = {
  isVisible: false,
  toggle: toggleStub,
  confirm: confirmStub
}

describe('ProfileForm', () => {

  let wrapper, instance;

  beforeEach(() => {
    wrapper = shallow(<ProfileForm {...someProps} />)
    instance = wrapper.instance();
  })

  // tests
  it('renders without crashing', async () => {
    wrapper.update();
  });

  it('renders correctly', async () => {
    const tree = await renderer
      .create(<ProfileForm />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('contains the correct elements', () => {

    ['Modal', 'ModalBody', 'ModalFooter', 'Form',{ color: "primary" }, { color: "primary" }].forEach((selector) => {
      expect(wrapper.find(selector).length).toEqual(1);
    });

    expect(wrapper.find('Input').length).toBe(3);
    expect(wrapper.find('Input').at(0).props().placeholder).toBe("Say something about yourself!");
    expect(wrapper.find('Input').at(1).props().placeholder).toBe("What is your area of expertise?")
    expect(wrapper.find('Input').at(2).props().placeholder).toBe("Where do you conduct your research?")

  });

  it('should fire the provided toggle when the cancel button is clicked', () => {

    wrapper.find({ color: "secondary" }).simulate('click');
    expect(toggleStub.mock.calls.length).toBe(1);

  })

  it('should have an empty profileInfo in state to start', () => { 
    expect(wrapper.state().profileInfo).toMatchObject({})
  })

  it('should save profileInfo in state when user types in info', () => { 
    let e = {
      target: {
        name: 'about',
        value: 'My name is mark!'
      }
    }
    
    wrapper.instance().handleChange(e)
    wrapper.update();
    expect(wrapper.state().profileInfo).toMatchObject({
      about: 'My name is mark!'
    });

  })

  it('should fire the provided this.props.confirm function when OK is clicked', () => {

    wrapper.find({color: 'primary'}).simulate('click');
    wrapper.update();

    expect(confirmStub.mock.calls.length).toBe(1)
  })

})