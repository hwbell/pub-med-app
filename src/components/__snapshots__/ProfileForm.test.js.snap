// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileForm renders correctly 1`] = `null`;
